Auxiliary flight control actuator handling rework Take2

sfuhrer opened the pull request:
Alternative to https://github.com/PX4/PX4-Autopilot/pull/20724, based on https://github.com/PX4/PX4-Autopilot/pull/20237

New features:

- enable flaperons (deflect ailerons down to act as flaps) and spoilerons (deflect ailerons up to act as spoilers) with control allocation (Fixes Control Allocator - Can't tie ailerons up (crow) to flap deployment #20694)
- add option to map manual spoiler setpoint to Aux1 or the flaps channel (for linked deployment of flaps+spoiler with single knob/button)
- enables landing gear logic for FW in Auto flight (retract after takeoff, deploy before landing)

Clean up:
- removes collective tilt, landing gear, flaps, spoilers and airbrakes from actuator_controls --> enables us to get rid of actuator_controls
- removes flaps/spoiler related fields from vehicle_attitdue_setpoint
- remove `FW_DTRIM_P_SPOIL`, `FW_DTRIM_P_FLAPS` from the rate controller as the trimming can now be configured in the allocator

How it works:
- add three new topics that replace current fields of actuator_controls, that are used to pass controller output to the allocation:
```
tiltrotor_extra_controls.msg
flaps_setpoint.msg
spoilers_setpoint.msg
```
![image](https://user-images.githubusercontent.com/26798987/213479994-a12de527-91c0-46ba-8416-301876128603.png)

- allow several modules to publish same topic, and control through flags e.g. spoiler_setpoint from FWAttitudeController if `!flag_control_auto_enabled`, or FWPositionController if flag_control_auto_enabled and in FW flight, or VTOLAttitudeController if in Auto hover and VTOL.
- manual flaps and spoiler control is done in FixedwingAttitudeController, in Auto flight in FixedwingPositionController 
- Control allocation side: same as in https://github.com/PX4/PX4-Autopilot/pull/20724, so: 
![image](https://user-images.githubusercontent.com/26798987/206231120-4b1612c1-0d36-4e98-9b8c-72af3b5bcadd.png)

[Link to video benchtesting flaps/spoilers](https://discord.com/channels/1022170275984457759/1022186266957201480/1050083696075620443)

Comparison to https://github.com/PX4/PX4-Autopilot/pull/20724:
+uses less custom topics
+no additional module needed

-multiple places to publish same topic instead of having a central place, with having to rely on flaps to turn on/off the publishing in the different controllers --> makes it more prone to errors and race conditions
-manual flaps and spoiler control is done in FixedwingAttitudeController, which is a bit of a misuse of an "attitude controller", and is inconsistent with the callback of the attitude controller (which is attitude estimate, which isn't necessary for controlling flaps/spoilers in full Manual mode)

Arguably this can be done nicer once we have clear distinct flight modes, though it then would ofc mean that we will have to copy some of the logic into several modes (e.g. landing gear should be handles the same for all Manual modes).

tstastny commented on src/modules/fw_pos_control_l1/fw_pos_control_l1_params.c:
```diff
@@ -1074,3 +1074,57 @@ PARAM_DEFINE_FLOAT(FW_WIND_ARSP_SC, 0.f);
+ */
+PARAM_DEFINE_FLOAT(FW_FLAPS_LND_SCL, 1.0f);
+
+/**
+ * Spoiler landing setting
+ *
+ * @unit norm
+ * @min 0.0
+ * @max 1.0
+ * @decimal 2
+ * @increment 0.01
+ * @group FW Attitude Control
+ */
+PARAM_DEFINE_FLOAT(FW_SPOILERS_LND, 0.f);
+
+/**
+ * Spoiler descend setting
+ *
+ * @unit norm
+ * @min 0.0
+ * @max 1.0
+ * @decimal 2
+ * @increment 0.01
+ * @group FW Attitude Control
+ */
+PARAM_DEFINE_FLOAT(FW_SPOILERS_DESC, 0.f);
```
also applies to spoilerons if enabled in mixer?

sfuhrer replied:
Yes, the controller doesn't care about how the "spoiler setpoint" is allocated, can be with a dedicated spoiler, spoilerons or a mix of both, what ever is configured in the allocations.

tstastny reviewed (changes requested):
Happy we can start cleaning these things up at least incrementally.

> enables us to get rid of actuator_controls
but it does not get rid of it yet in this PR - what are the lingering reasons we need the actuator_controls msg?

> Arguably this can be done nicer once we have clear distinct flight modes,  though it then would ofc mean that we will have to copy some of the logic into several modes
a bit of copy paste is ok imo, much rather have clean interfaces and modularity.

sfuhrer commented on msg/ActuatorControls.msg:
```diff
@@ -5,13 +5,8 @@ uint8 INDEX_ROLL = 0
 uint8 INDEX_PITCH = 1
 uint8 INDEX_YAW = 2
 uint8 INDEX_THROTTLE = 3
```
Will anyway be further changed in https://github.com/PX4/PX4-Autopilot/pull/20837

sfuhrer commented on src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessControlSurfaces.cpp:
```diff
@@ -151,32 +161,22 @@ bool ActuatorEffectivenessControlSurfaces::addActuators(Configuration &configura
+		ActuatorVector &actuator_sp) const
+{
+	for (int i = 0; i < _count; ++i) {
+		// map [0, 1] to [-1, 1]
+		// TODO: this currently only works for spoilerons, not dedicated spoilers
+		actuator_sp(i + first_actuator_idx) += spoilers_control * _params[i].scale_spoiler;
```
No, it doesn't, it's a wrong comment. Anyway, the core issue is that we need to find a unified way to set one-way control surfaces (currently all control surfaces need range of [-1, 1], while it should be [0, 1]).

sfuhrer commented on src/modules/fw_att_control/FixedwingAttitudeControl.cpp:
```diff
@@ -277,13 +277,9 @@ void FixedwingAttitudeControl::Run()
 		bool wheel_control = false;
 
-		if (_param_fw_w_en.get() && _att_sp.fw_control_yaw_wheel) {
+		if (_param_fw_w_en.get() && _att_sp.fw_control_yaw_wheel && _vcontrol_mode.flag_control_auto_enabled) {
```
Didn't try it out but sounds feasible to me. 

>Then we could maybe extract wheel control out of attitude control entirely

Move it into it's own module you mean?

tstastny replied:
> Move it into it's own module you mean?
Yeah - or at least it's own class to begin with. Not necessary in this PR necessarily - could be follow up post some more discussion.

sfuhrer commented:
Rebased and updated description with a diagram. @dagar could you take a closer look here?

sfuhrer commented:
Rebased on latest main. 
@ryanjAA fyi, in case you want to flight test including the FOH fix.

dagar commented on msg/FlapsSetpoint.msg:
```diff
@@ -0,0 +1,3 @@
+uint64 timestamp        		# time since system start (microseconds)
```
We should merged these common types (the actual .msg) and then have different topics for ORB_ID(flaps_setpoint), ORB_ID(spoilers_setpoint), etc.

Something like ActuatorSetpoint.msg or NormalizedActuatorSetpoint.msg?

sfuhrer replied:
But it's not directly an actuator setpoint - e.g. spoiler_setpoint.normalized_setpoint can be mapped to ailerons. Calling it "Actuator" something sounds confusion to me.
Agree that we could combine the Flaps and Spoiler "controls" topics, just need to find a good name. AerodynamicDevicesConfiguration.msg, with FlapsConfiguration and SpoilerConfiguration? Or Configuration replaced with Setpoint? Or back to Controls?

dagar replied:
Or even just NormalizedSetpoint.msg could be fine, the main point is to share these common data types (and get away from .msg === topic).

sfuhrer replied:
Will make the change and rebase on latest main after the L1 purge is through.

ryanjAA replied:
We'll test when it's done (flying tomorrow but pretty much every few days so no rush for tomorrow).

dagar commented:
Why handle this in `fw_att_control` vs `fw_rate_control`? At least if it was at the lowest level that would be a better answer for manual mode, etc.

Won't this be kind of bad if there's any flap or spoiler deployment and you switch between any altitude/position mode or simpler attitude/rate/manual mode?

dagar commented:
>Won't this be kind of bad if there's any flap or spoiler deployment and you switch between any altitude/position mode or simpler attitude/rate/manual mode?

Push the slew for flaps/spoilers into control allocator (instead of fw_pos_ctrl, etc)?

dagar commented on src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessTiltrotorVTOL.cpp:
```diff
@@ -133,16 +135,12 @@ void ActuatorEffectivenessTiltrotorVTOL::updateSetpoint(const matrix::Vector<flo
-			if (_actuator_controls_0_sub.copy(&actuator_controls_0)) {
 				for (int i = 0; i < _first_tilt_idx; ++i) {
-					actuator_sp(i) = actuator_controls_0.control[actuator_controls_s::INDEX_THROTTLE];
+					actuator_sp(i) = tiltrotor_extra_controls.collective_thrust_normalized_setpoint;
```
This can't come through a regular thrust setpoint now?

sfuhrer replied:
We would need to hack the API, as a tilted thrust doesn't fit in only one of vehicle_thrust_setpoint.[x]. The clean thing to do would be to give the thrust vector along in vector form, but for that we would need to change quite a bit in tiltrotor.cpp and ActuatorEffectivenessTiltrotor. I would prefer to do this in a follow up as a stand alone task, as it's hard to asses the complexity.

tstastny commented:
>Push the slew for flaps/spoilers into control allocator (instead of fw_pos_ctrl, etc)?

this depends on if we want different flaps/spoiler deployment behavior in different modes. maybe in e.g. manual we dont want a slew or filter at all. i guess baseline there needs to be some small actuator slew in general on all control surfaces (just to avoid burning them out) which could go in on the allocator level, but for heavier filters like the alpha ones we have in most modes now, probably should be specific to the mode desires, no?

dagar commented:
> > Push the slew for flaps/spoilers into control allocator (instead of fw_pos_ctrl, etc)?
> 
> this depends on if we want different flaps/spoiler deployment behavior in different modes. maybe in e.g. manual we dont want a slew or filter at all. i guess baseline there needs to be some small actuator slew in general on all control surfaces (just to avoid burning them out) which could go in on the allocator level, but for heavier filters like the alpha ones we have in most modes now, probably should be specific to the mode desires, no?

@tstastny I was specifically thinking of it as a bit of a hack to minimize any nasty transitions between high level and low level flight modes if flaps or spoilers are deployed at the time of the switch.

dagar commented:
> > @tstastny I was specifically thinking of it as a bit of a hack to minimize any nasty transitions between high level and low level flight modes if flaps or spoilers are deployed at the time of the switch.
> 
> i guess this would need some mode transition logic then -- since we don't have feedback on the control surface state, would need to initialized filters at last control command (which we should have). this would handle it for low -> high level mode transitions. But for high -> low level mode transitions, in worst case would be the barebones slew only that protects the setpoint. But that would maybe be ok IMO?

This is why I originally proposed having a standalone little module own the flaps, etc. 

If a conservative slew at the control allocator level is tolerable I think it could be fine, allowing us to keep the controller usage simple skipping extra transition logic, filter init, etc. I think the main downside is that it might feel a bit weird when your manual flaps control is so slow, or we could even consider it a safety feature.

sfuhrer commented:
> Why handle this in `fw_att_control` vs `fw_rate_control`? At least if it was at the lowest level that would be a better answer for manual mode, etc.

The rate controller is as wrong as the attitude controller. We can also move it there, but I would just do it in the same module as the wheel controller. Once we have nice separated flight modes I hope to be able to move this stuff there.

> Won't this be kind of bad if there's any flap or spoiler deployment and you switch between any altitude/position mode or simpler attitude/rate/manual mode?

Switching between any manual modes won't change anything (always directly listening to the RC switch). Switching from Auto to Manual will then ofc involve switches that we don't cover with slew rates, unless the slew rate is on the actuator (only works for dedicated actuator, not if ailerons are used for it). But how often do you switch between these modes while you have any flap/spoiler deployment..likely the most critical during manual landing aborts. 

>Push the slew for flaps/spoilers into control allocator (instead of fw_pos_ctrl, etc)?
The actuators already have a slew rate, but this doesn't work for flaperons (as we really don't want a slew rate on the ailerons).

tstastny commented:
> Once we have nice separated flight modes I hope to be able to move this stuff there.

right - ideally we'd have some library that takes flaps general command within mode and translates it (through whatever config the user sets) to direct actuator commands which bypass control allocation - this would mean that we have clean state based control api (no "exotic" setpoints necessary), but still allow modes to do what they will with flaps in an easy to work with manner for the writer of said mode. Also means any given mode can decide what filtering to use etc.. and no need for independent module. But this can come later. I guess closest thing to this now would be putting all flaps deployment knowledge into fw position control module -- so at least then we remove them from attitude and rate controls.

> Yes, having the ability to slew rate flaps/spoilers over manual<.>auto mode transitions is for me the outstanding feature that we can't solve cleanly without a stand alone module.

I think it can be - just needs to subscribe to the flaps setpoint topic and initialize with the last state, no? (not saying this needs to happen in this pr)

sfuhrer commented on src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessFixedWing.cpp:
```diff
@@ -61,21 +61,20 @@ ActuatorEffectivenessFixedWing::getEffectivenessMatrix(Configuration &configurat
-	}
+	_flaps_setpoint_sub.copy(&flaps_setpoint);
+
+	_control_surfaces.applyFlaps(flaps_setpoint.normalized_setpoint, _first_control_surface_idx, dt, actuator_sp);
```
prob needs guards against NaN (when there is no flaps_setpoint message or the field is NaN?)

sfuhrer commented:
> I think it can be - just needs to subscribe to the flaps setpoint topic and initialize with the last state, no? (not saying this needs to happen in this pr)

Right, if we do it in the CA then the flaps/spoilers slew rate limiting can be done consistently over all flight modes (and also on mode changes). 
Pushed a proposal in 0e724d7f1dcf0dc3739fb4d312fddc4d1e16c4e4. @bkueng curious for your inputs on it. Instead of adding more stuff into `updateSetpoint()` I opted for adding a new method `allocateAuxilaryControls()` that updates the slew rated values for the flaps and spoilers state, plus then calls `_control_surfaces.applySpoilers()` for applying them. That way `updateSetpoint()` can maybe be rather kept for the main actuation. Let me know what you think.

bkueng reviewed (commented):
> -multiple places to publish same topic instead of having a central place, with having to rely on flaps to turn on/off the publishing in the different controllers --> makes it more prone to errors and race conditions
-manual flaps and spoiler control is done in FixedwingAttitudeController, which is a bit of a misuse of an "attitude controller", and is inconsistent with the callback of the attitude controller (which is attitude estimate, which isn't necessary for controlling flaps/spoilers in full Manual mode)

At least for later on I was thinking along these lines:
- mode tells the mode manager/base if it wants to control any of these
- if so, mode publishes the setpoint
- if not, manager/base publishes based on RC/config/...

sfuhrer commented on src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessControlSurfaces.cpp:
```diff
@@ -151,32 +164,25 @@ bool ActuatorEffectivenessControlSurfaces::addActuators(Configuration &configura
-			break;
+	for (int i = 0; i < _count; ++i) {
+		// map [0, 1] to [-1, 1]
+		// TODO: this currently only works for dedicated flaps, not flaperons
```
Correct me if I'm wrong, but currently to get a control surfaces to move to PWM_MIN you have to set it's actuator_sp to -1. There is no CENTER setting that would allow you to say "map actuator_sp 0 to XX PWM". 
The flaps_setpoint is a signed desired configuration (0 deployment to full deployment), which means I have to map [0,1] to [-1, 1] to achieve full range on dedicated flaps. When I though want to allocate the flaps configuration to ailerons, which can move up and down, the [0,1] could be directly applied.

bkueng replied:
Right, we probably don't get around to make some kind of distinction for these 2 use-cases.

sfuhrer replied:
Yes we can discuss that in a follow up.

sfuhrer commented on src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessFixedWing.cpp:
```diff
@@ -61,18 +61,20 @@ ActuatorEffectivenessFixedWing::getEffectivenessMatrix(Configuration &configurat
-	actuator_controls_s actuator_controls_0;
+	normalized_unsigned_setpoint_s flaps_setpoint;
+
+	_flaps_setpoint_sub.copy(&flaps_setpoint);
```
Hmm no, applyFlaps() needs to be called always as it includes the slew rate, which also has to be updated if there is no new flap setpoint coming in.

bkueng replied:
`copy` will still do that. It only returns false if the topic does not exist.

sfuhrer replied:
Ah yes ofc, will adapt it.

sfuhrer commented on src/modules/control_allocator/ActuatorEffectiveness/ActuatorEffectivenessStandardVTOL.cpp:
```diff
@@ -63,22 +63,22 @@ ActuatorEffectivenessStandardVTOL::getEffectivenessMatrix(Configuration &configu
+void ActuatorEffectivenessStandardVTOL::allocateAuxilaryControls(const float dt, ActuatorVector &actuator_sp)
 {
 	// apply flaps
-	if (matrix_index == 1) {
```
Added it back.

sfuhrer commented on src/modules/control_allocator/module.yaml:
```diff
@@ -606,6 +694,20 @@ mixer:
               - { 'min': 0.0, 'max': 1.0, 'default': 0.5 } # pitch
               - { 'min': -1.0, 'max': 0.0, 'default': -0.5 } # yaw
 
+      - select_identifier: 'servo-type-tailsitter' #
+        apply_identifiers: ['servo-torque-roll-tailsitter', 'servo-torque-pitch-tailsitter', 'servo-torque-yaw-tailsitter']
+        items:
+        # Convention: horizontal surfaces: up=positive, vertical: right=positive, mixed: up=positive.
+        # No rules for control surfaces on tailsitter beside elevons.
```
right, removed it.

sfuhrer commented on src/modules/fw_path_navigation/CMakeLists.txt:
```diff
@@ -43,6 +43,8 @@ px4_add_module(
 		launchdetection
     		npfg
+		SlewRate
+    npfg
```
also SlewRate was duplicated, removed them both.

sfuhrer commented:
I've rebased, cleaned up the commits and fixed some things:
- move the manual flaps/spoilers handling into FW rate controller (@dagar as discussed)
- addressed your comments inside CA @bkueng https://github.com/PX4/PX4-Autopilot/pull/20815#discussion_r1119722887
- made sure that there is always a flaps/spoiler setting published in all modes and VTOL/FW

Then flight tested it on a Tiltrotor VTOL and bench tested on a plane with flaps, landing gear and nose wheel, no issues found:
https://review.px4.io/plot_app?log=d45309fe-2b14-4564-8e93-45c3c1cd7167

I hope this is now ready to get in!

sfuhrer commented:
Had to rebase to resolve conflict that came in with https://github.com/PX4/PX4-Autopilot/pull/21192 in FlightTaskDescend.

sfuhrer commented on src/modules/flight_mode_manager/tasks/Descend/FlightTaskDescend.cpp:
```diff
@@ -36,6 +36,13 @@
 
 #include "FlightTaskDescend.hpp"
 
+bool FlightTaskDescend::activate(const trajectory_setpoint_s &last_setpoint)
```
Had to rebase to resolve conflict that came in with https://github.com/PX4/PX4-Autopilot/pull/21192. There the activate() was removed, so I add it back here.

sfuhrer commented:
Rebased on main to solve flash issues.

ryanjAA commented:
Will be flying the latest of this tomorrow FYI
